src/boot_O2_g3/fault.c

```diff
@@ -564,354 +695,440 @@ void Fault_DrawMemDump(u32 pc, u32 sp, u32 unk0, u32 unk1) {
+ * possible to backtrace through an eret as an interrupt can occur at any time, so
+ * there is no choice but to give up here. For j instructions, they can be followed
+ * and the backtrace may continue as normal.
+ */
+void Fault_WalkStack(uintptr_t* spPtr, uintptr_t* pcPtr, uintptr_t* raPtr) {
     uintptr_t sp = *spPtr;
     uintptr_t pc = *pcPtr;
     uintptr_t ra = *raPtr;
-    u32 lastOpc;
-    u16 opcHi;
-    s16 opcLo;
+    u32 lastInsn;
+    u16 insnHi;
+    s16 insnLo;
     u32 imm;
 
-    if (sp & 3 || sp < 0x80000000 || sp >= 0xC0000000 || ra & 3 || ra < 0x80000000 || ra >= 0xC0000000) {
+    if ((sp % 4 != 0) || (sp < K0BASE) || (sp >= K2BASE) || (ra % 4 != 0) || (ra < K0BASE) || (ra >= K2BASE)) {
         *spPtr = 0;
         *pcPtr = 0;
         *raPtr = 0;
         return;
     }
 
-    if (pc & 3 || pc < 0x80000000 || pc >= 0xC0000000) {
+    if ((pc % 4 != 0) || (pc < K0BASE) || (pc >= K2BASE)) {
         *pcPtr = ra;
         return;
     }
 
-    lastOpc = 0;
+    lastInsn = 0;
     while (true) {
-        opcHi = *(uintptr_t*)pc >> 16;
-        opcLo = *(uintptr_t*)pc & 0xFFFF;
-        imm = opcLo;
+        insnHi = *(uintptr_t*)pc >> 16;
+        insnLo = *(uintptr_t*)pc & 0xFFFF;
+        imm = insnLo;
 
-        if (opcHi == 0x8FBF) {
+        if (insnHi == 0x8FBF) {
+            // lw $ra, <imm>($sp)
+            // read return address saved on the stack
             ra = *(uintptr_t*)(sp + imm);
-        } else if (opcHi == 0x27BD) {
+        } else if (insnHi == 0x27BD) {
+            // addiu $sp, $sp, <imm>
+            // stack pointer increment or decrement
             sp += imm;
         } else if (*(uintptr_t*)pc == 0x42000018) {
+            // eret
+            // cannot backtrace through an eret, give up
             sp = 0;
             pc = 0;
             ra = 0;
-            goto end;
+            goto done;
         }
-        if (lastOpc == 0x3E00008) {
+        if (lastInsn == 0x3E00008) {
+            // jr $ra
+            // return to previous function
             pc = ra;
-            goto end;
-        } else if ((lastOpc >> 26) == 2) {
-            pc = pc >> 28 << 28 | lastOpc << 6 >> 4;
-            goto end;
+            goto done;
+        } else if (lastInsn >> 26 == 2) {
+            // j <target>
+            // extract jump target
+            pc = (pc >> 28 << 28) | (lastInsn << 6 >> 4);
+            goto done;
         }
-        lastOpc = *(uintptr_t*)pc;
-        pc += 4;
+
+        lastInsn = *(uintptr_t*)pc;
+        pc += sizeof(u32);
     }
 
-end:
+done:
     *spPtr = sp;
     *pcPtr = pc;
     *raPtr = ra;
 }
 
+/**
+ * Draws the stack trace page contents for the specified thread
+ */
 void Fault_DrawStackTrace(OSThread* t, u32 flags) {
-    s32 y;
-    u32 sp;
-    u32 ra;
-    u32 pc;
+    s32 line;
+    uintptr_t sp = t->context.sp;
+    uintptr_t ra = t->context.ra;
+    uintptr_t pc = t->context.pc;
     u32 pad;
```

Comment:
```suggestion
    s32 pad;
```
Also can it go below `addr`?